src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6271,229 +6271,205 @@ private TypeSyntax ParsePointerTypeMods(TypeSyntax type)
                 StackGuard.EnsureSufficientExecutionStack(_recursionDepth);
 
-                if (this.IsIncrementalAndFactoryContextMatches && this.CurrentNode is CSharp.Syntax.StatementSyntax)
-                {
-                    return (StatementSyntax)this.EatNode();
-                }
-
-                // First, try to parse as a non-declaration statement. If the statement is a single
-                // expression then we only allow legal expression statements. (That is, "new C();",
-                // "C();", "x = y;" and so on.)
+                StatementSyntax result;
 
-                StatementSyntax result = ParseStatementNoDeclaration(allowAnyExpression: false);
-                if (result != null)
+                // Main switch to handle processing almost any statement.
+                switch (this.CurrentToken.Kind)
                 {
-                    return result;
+                    case SyntaxKind.FixedKeyword:
+                        return this.ParseFixedStatement();
+                    case SyntaxKind.BreakKeyword:
+                        return this.ParseBreakStatement();
+                    case SyntaxKind.ContinueKeyword:
+                        return this.ParseContinueStatement();
+                    case SyntaxKind.TryKeyword:
+                    case SyntaxKind.CatchKeyword:
+                    case SyntaxKind.FinallyKeyword:
+                        return this.ParseTryStatement();
+                    case SyntaxKind.CheckedKeyword:
+                    case SyntaxKind.UncheckedKeyword:
+                        return this.ParseCheckedStatement();
+                    case SyntaxKind.DoKeyword:
+                        return this.ParseDoStatement();
+                    case SyntaxKind.ForKeyword:
+                        return this.ParseForOrForEachStatement();
+                    case SyntaxKind.ForEachKeyword:
+                        return this.ParseForEachStatement(awaitTokenOpt: default);
+                    case SyntaxKind.GotoKeyword:
+                        return this.ParseGotoStatement();
+                    case SyntaxKind.IfKeyword:
+                        return this.ParseIfStatement();
+                    case SyntaxKind.ElseKeyword:
+                        // Including 'else' keyword to handle 'else without if' error cases 
+                        return this.ParseMisplacedElse();
+                    case SyntaxKind.LockKeyword:
+                        return this.ParseLockStatement();
+                    case SyntaxKind.ReturnKeyword:
+                        return this.ParseReturnStatement();
+                    case SyntaxKind.SwitchKeyword:
+                        return this.ParseSwitchStatement();
+                    case SyntaxKind.ThrowKeyword:
+                        return this.ParseThrowStatement();
+                    case SyntaxKind.UnsafeKeyword:
+                        result = TryParseStatementStartingWithUnsafe();
```

Comment:
just curious, did you measure if there was any difference in declaring a var in each switch label versus declaring it before the switch?

Reply:
i didn't measure. it would likely be easier to test taht in your branch once this goes in :)  would be fine either way (if they're the same).